### PR TITLE
Improve support for opening multiple clients at once

### PR DESCRIPTION
**Pull request by xPaw**
Improve support for opening multiple clients at once

- Synchronize unread counter with the server
- Fix unread marker on no attached clients #561
- Increase unread counter for server messages
- Keep track of all the open channels across all client sockets

### CODE REVIEW

**Comment by xPaw** [comment]
> 08:48 astorije  Why doesn't this solve the problem?
09:08 +xPaw  Because all you do is mark which client reads which channel (same exact problem as  now), it will still get out of sync between channels and page reloads
09:10 +xPaw  I'm thinking unread counter has to be global,  but marking open channels per client (keep the last open channel for page reloads), then the desync problem should be solved
09:11 +xPaw  Then when one client opens a channel, tell other clients to mark it as read

File: client/js/lounge.js [hunk 1]
```diff
@@ -170,6 +170,10 @@ $(function() {
 		$("#sign-in").remove();
 
 		var id = data.active;
+
+		// Set it right away before triggering a click so that "open" event is not emitted
```
Comment (xPaw) [inline client/js/lounge.js]: Maybe this isn't actually a good idea, as loading the page *should (?)* mark the channel as read.

Comment (MaxLeiter) [inline client/js/lounge.js]: Marking the page as read removes the unread marker, right? If so, I think this is fine and the page should be left as unread.

Comment (astorije) [inline client/js/lounge.js]: Yes, when loading the page, the active channel shouldn't be marked as read to give a chance to the user to catch up on new messages.

Actually, all that chain of PRs (#561, #613, this one, #635) was triggered by trying to not mark channel as read on initial load :D

**Comment by astorije** [comment]
@xPaw, I'm unclear if you want to make changes or not, feel free to assign me to that PR when you think it's ready to go.

File: client/js/lounge.js (hunk 1 above)
Comment (xPaw) [inline client/js/lounge.js]: It's actually quite silly this way. If you open the page and `#channel1` is loaded, switch to another channel and then reload the page, `#channel1` will display the previous unread counter as it won't be reset.

**Comment by astorije** [comment]
Testing this, I noticed an out-of-sync with unread counters on multiple clients:

counter-not-synced

Note that `[12]` does not clear on second client until a new message gets posted.

File: client/js/lounge.js [hunk 2]
```diff
@@ -181,6 +181,13 @@ $(function() {
 		}
 	});
 
+	socket.on("open", function(id) {
+		// Another client opened the channel, clear the unread counter
+		sidebar.find("[data-id='" + id + "'] .badge")
+			.removeClass("highlight")
+			.empty();
+	});
+
```
Comment (xPaw) [inline client/js/lounge.js]: What do you mean?

**Comment by xPaw** [comment]
@astorije fixed.

File: client/js/lounge.js (hunk 2 above)
Comment (astorije) [inline client/js/lounge.js]: I mean that this clears unread counter for channels opened in other clients, do we need to move the unread **marker** for these channels as well?

File: client/js/lounge.js [hunk 3]
```diff
@@ -840,7 +847,6 @@ $(function() {
 		self.addClass("active")
 			.find(".badge")
 			.removeClass("highlight")
-			.data("count", 0)
```
Comment (astorije) [inline client/js/lounge.js]: Why are we getting rid of this?

**Comment by xPaw** [comment]
There's no longer a need for it, as it only ever updates the badge from the
server now.

File: client/js/lounge.js (hunk 3 above)
Comment (xPaw) [inline client/js/lounge.js]: There's no longer a need for it, as it only ever updates the badge from the
server now.

**Pull request merged**